### PR TITLE
Attributesprocessor

### PR DESCRIPTION
**Pull request by bogdandrutu**
Fixes in `service/builder/` are a bit hacky until we change example factories to use the new data-model.

### CODE REVIEW

File: internal/data/common.go [hunk 1]
```diff
@@ -109,30 +117,52 @@ func (a AttributeValue) BoolVal() bool {
 }
 
 func (a AttributeValue) SetString(v string) {
-	a.orig.Type = otlpcommon.AttributeKeyValue_STRING
+	a.setType(otlpcommon.AttributeKeyValue_STRING)
```
Comment (tigrannajaryan) [inline internal/data/common.go:120]: 👍 
Otherwise we would keep all fields around and pollute the wire when marshaling.

**Comment by bogdandrutu** [comment]
@tigrannajaryan PTAL, now all the dependencies were merged.

**Comment by bogdandrutu** [comment]
A lot of the changes are in tests unfortunately...

File: processor/attributesprocessor/attributes.go [hunk 2]
```diff
@@ -126,39 +101,61 @@ func (a *attributesProcessor) Shutdown(context.Context) error {
...
 
-func setAttribute(action attributeAction, attributesMap map[string]*tracepb.AttributeValue) {
+func getAttributeValue(action attributeAction, attrs data.AttributeMap) (data.AttributeValue, bool) {
 	// Set the key with a value from the configuration.
-	if action.AttributeValue != nil {
-		attributesMap[action.Key] = action.AttributeValue
-	} else if value, fromAttributeExists := attributesMap[action.FromAttribute]; fromAttributeExists {
-		// Set the key with a value from another attribute, if it exists.
-		attributesMap[action.Key] = value
+	if !action.AttributeValue.IsNil() {
```
Comment (tigrannajaryan) [inline processor/attributesprocessor/attributes.go:149]: This needs to be done differently after merging https://github.com/open-telemetry/opentelemetry-collector/pull/809/files

File: processor/attributesprocessor/attributes.go [hunk 3]
```diff
@@ -126,39 +101,61 @@ func (a *attributesProcessor) Shutdown(context.Context) error {
...
+			}
+			attrs.Upsert(action.Key, av)
+		case HASH:
+			hashAttribute(action, attrs)
+		}
+	}
 }
 
-func setAttribute(action attributeAction, attributesMap map[string]*tracepb.AttributeValue) {
+func getAttributeValue(action attributeAction, attrs data.AttributeMap) (data.AttributeValue, bool) {
```
Comment (tigrannajaryan) [inline processor/attributesprocessor/attributes.go:147]:
```suggestion
func getSourceAttributeValue(action attributeAction, attrs data.AttributeMap) (data.AttributeValue, bool) {
```

File: processor/attributesprocessor/attributes.go (hunk 2 above)
Comment (bogdandrutu) [reply processor/attributesprocessor/attributes.go]: Done.

File: processor/attributesprocessor/attributes.go (hunk 3 above)
Comment (bogdandrutu) [reply processor/attributesprocessor/attributes.go:147]: Done.

File: processor/attributesprocessor/attribute_hasher.go [hunk 4]
```diff
@@ -38,23 +38,23 @@ var (
 // hashed version of the attribute. In practice, this would mostly be used
 // for string attributes but we support all types for completeness/correctness
 // and eliminate any surprises.
-func SHA1AttributeHahser(attr *tracepb.AttributeValue) *tracepb.AttributeValue {
+func SHA1AttributeHahser(attr data.AttributeValue) {
```
Comment (pjanotti) [inline processor/attributesprocessor/attribute_hasher.go:41]:
```suggestion
func SHA1AttributeHasher(attr data.AttributeValue) {
```

File: processor/attributesprocessor/attributes.go [hunk 5]
```diff
@@ -126,39 +101,61 @@ func (a *attributesProcessor) Shutdown(context.Context) error {
...
+		// TODO https://github.com/open-telemetry/opentelemetry-collector/issues/296
+		// Do benchmark testing between having action be of type string vs integer.
+		// The reason is attributes processor will most likely be commonly used
+		// and could impact performance.
+		switch action.Action {
+		case DELETE:
+			attrs.Delete(action.Key)
+		case INSERT:
+			av, found := getSourceAttributeValue(action, attrs)
+			if !found {
```
Comment (pjanotti) [inline processor/attributesprocessor/attributes.go:125]: if `found` continue

File: processor/attributesprocessor/attributes.go [hunk 6]
```diff
@@ -126,39 +101,61 @@ func (a *attributesProcessor) Shutdown(context.Context) error {
...
+				continue
+			}
+			attrs.Insert(action.Key, av)
```
Comment (pjanotti) [inline processor/attributesprocessor/attributes.go:128]: hum, the name of the method here implies that it takes care if the attribute exists or not, is that the case?

Comment (bogdandrutu) [reply processor/attributesprocessor/attributes.go:128]: Correct. I think you are confused by `getSourceAttributeValue`

File: processor/attributesprocessor/attributes.go (hunk 5 above)
Comment (bogdandrutu) [reply processor/attributesprocessor/attributes.go:125]: getSourceAttributeValue determines what value should be set for the "key" in the config:
* If the user specified a value in the cofig use that
* If the user specified a value from another attribute in the map, retrieve that.

Comment (bogdandrutu) [reply processor/attributesprocessor/attributes.go:125]: So if we don't found a value to apply in the operation we don't have anything to do.

File: processor/attributesprocessor/attribute_hasher.go (hunk 4 above)
Comment (bogdandrutu) [reply processor/attributesprocessor/attribute_hasher.go:41]: done.

**Pull request merged**